google-cloud-storage/lib/google/cloud/storage/policy.rb

```diff
@@ -162,43 +186,76 @@ def role role_name
         end
 
         ##
-        # Returns a deep copy of the policy.
+        # Returns a deep copy of the policy. The service object referenced by
+        # the policy is not duplicated.
         #
         # @return [Policy]
         #
         def deep_dup
           dup.tap do |p|
-            roles_dup = p.roles.each_with_object({}) do |(k, v), memo|
-              memo[k] = v.dup rescue value
+            gapi_dup = @gapi.dup
+            bindings_dup =  @gapi.bindings.map do |binding|
+              binding_dup = binding.dup
+              binding_dup.members = binding.members.dup
+              binding_dup
             end
-            p.instance_variable_set "@roles", roles_dup
+            gapi_dup.bindings = bindings_dup
+            p.gapi = gapi_dup
           end
         end
 
         ##
+        # Reloads the policy with current data from the Cloud IAM service.
+        #
+        # @example Retrieve the latest policy from the service:
+        #   require "google/cloud/storage"
+        #
+        #   storage = Google::Cloud::Storage.new
+        #
+        #   bucket = storage.bucket "my-todo-app"
+        #
+        #   policy = bucket.policy force: true # API call
+        #   policy_2 = policy.reload! # API call
+        #
+        def reload!
+          ensure_service!
+          bucket_name = resource_id.split("/").last
+          # TODO: replace line above with regex supporting generation
+          # TODO: conditional for get_object_policy
+          @gapi = service.get_bucket_policy bucket_name
+        end
+        alias_method :refresh!, :reload!
+
+        ##
         # @private Convert the Policy to a
         # Google::Apis::StorageV1::Policy.
         def to_gapi
-          Google::Apis::StorageV1::Policy.new(
-            etag: etag,
-            bindings: roles.keys.map do |role_name|
-              next if roles[role_name].empty?
-              Google::Apis::StorageV1::Policy::Binding.new(
-                role: role_name,
-                members: roles[role_name]
-              )
-            end
-          )
+          @gapi.bindings = roles.keys.map do |role_name|
+            next if roles[role_name].empty?
+            Google::Apis::StorageV1::Policy::Binding.new(
+              role: role_name,
+              members: roles[role_name]
+            )
+          end
+          @gapi
         end
 
         ##
         # @private New Policy from a
         # Google::Apis::StorageV1::Policy object.
-        def self.from_gapi gapi
-          roles = gapi.bindings.each_with_object({}) do |binding, memo|
-            memo[binding.role] = binding.members.to_a
+        def self.from_gapi gapi, conn
```

Comment:
The name of the second argument should probably be `service` instead of `conn`. I realize that this was missed in other code when migrating from Google API Client 0.8 to 0.9, but we shouldn't propagate that mistake.